pkg/utils/mainhelper/boot.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package mainhelper provides helper functions for common boilerplate code in
+// writing a main function such as setting up kube informers.
+package mainhelper
+
+import (
+	"context"
+	"log"
+	"net/http"
+
+	"github.com/google/knative-gcp/pkg/observability"
+	"github.com/kelseyhightower/envconfig"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/client-go/kubernetes"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/injection/sharedmain"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/profiling"
+)
+
+// BootRes holds a collection of objects after booting for convenient access by
+// other custom logic in the main function.
+type BootRes struct {
+	Ctx        context.Context
```

Comment:
context should always be threaded through directly. I would make this a separate return result and also make context the first argument to boot rather than an option.

Reply:
https://github.com/knative/pkg/blob/master/injection/sharedmain/main.go#L129

Reply:
I prefer having two function definitions so that ctx can be passed directly as an argument rather than through an option. This is mostly a nit, however, so feel free to merge as is if you prefer.

Reply:
I don't see a use case to pass a context. The sharedmain.MainWithContext is not called by anything either as far as I can tell. I can either remove it or keep it as an option. Just leaving it as an option for now.